### PR TITLE
Print Python Definitions with Inline Definitions for Tensor Dimensions

### PR DESCRIPTION
**Pull request by kevinstephano**
# Summary

**If the user can already use inline definitions in a `FusionDefinition`, what is the purpose of this PR?**
When users print definition, as happens when an error occurs, the reproduced definition did not include inline definitions meaning that repro scripts were much longer than the original definition and, therefore, hard to read!

* Adds inline definition printing for `Vector` and `Scalar`s used in the following operations
  * `reshape`
  * `broadcast_in_dim`
* Adds an `NVFUSER_DISABLE` option `python_inline_definitions` to toggle off inline definitions for debugging.
* The Python Frontend ends `State` objects were extended to include an `optional` `inline_def_record` private data member to enable printing of inline definitions.  This private state is not serialized!
* The Fusion Record's base class `RecordFunctor` includes an `inline_def` data member to enable printing the record properly when inline.

# Serialization/Deserialization Notes

`FusionDefinition`s and the `FusionCache` are both comprised of Fusion Records.  However, only `FusionDefinition`s include inline definitions.  Therefore, serialization does not include inline definition state.  The reason for this is that the cache does not understand what a `FusionDefinition` is, and therefore, the state required to reproduce the inline definition also does not exist.  

For inline definitions to work, each piece of state for a Fusion Record requires a pointer to its parent record.  In the case of the FusionCache, this relationship is not maintained therefore it does not make sense for deserialization to recreate these relationships because it only deserializes the `FusionCache` and not the corresponding `FusionDefinition`s.

# Examples

## A test
```python
import torch
from nvfuser import FusionDefinition, DataType

inputs = [torch.randn((24, 128, 25, 32), dtype=torch.float32, device="cuda:0")]

def fusion_func(fd: FusionDefinition) -> None:
    T18 = fd.define_tensor(
        shape=[-1, -1, -1, -1],
        contiguity=[True, True, True, True],
        dtype=DataType.Float,
        is_cpu=False,
    )
    T97 = fd.ops.reshape(T18, new_shape=[12, 128, 25, 32, 2])
    T98 = fd.ops.slice(
        T97,
        start_indices=[0, 0, 0, 0, 0],
        end_indices=[12, 128, 25, 32, 1],
        strides=[1, 1, 1, 1, 1],
    )
    T89 = fd.ops.sum(T98, dims=[4], keepdim=False, dtype=DataType.Null)
    fd.add_output(T89)

with FusionDefinition() as fd:
    fusion_func(fd)

for _ in range(5):
    out = fd.execute(inputs)
```

## Without Inline Definitions
Command Line:
```
NVFUSER_DISABLE=python_inline_definitions NVFUSER_DUMP=python_definition python test.py
```
Output:
```python
def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1, -1, -1, -1], contiguity=[True, True, True, True], dtype=DataType.Float, is_cpu=False)
    S1 = fd.define_scalar(12, dtype=DataType.Int)
    S2 = fd.define_scalar(128, dtype=DataType.Int)
    S3 = fd.define_scalar(25, dtype=DataType.Int)
    S4 = fd.define_scalar(32, dtype=DataType.Int)
    S5 = fd.define_scalar(2, dtype=DataType.Int)
    V6 = fd.define_vector([S1, S2, S3, S4, S5], dtype=DataType.Int)
    T7 = fd.ops.reshape(T0, new_shape=V6)
    T8 = fd.ops.slice(T7, start_indices=[0, 0, 0, 0, 0], end_indices=[12, 128, 25, 32, 1], strides=[1, 1, 1, 1, 1])
    T9 = fd.ops.sum(T8, dims=[4], keepdim=False, dtype=DataType.Null)
    fd.add_output(T9)
```
## With Inline Definitions
Command LIne:
```
NVFUSER_DUMP=python_definition python test.py
```
Output:
```python
def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1, -1, -1, -1], contiguity=[True, True, True, True], dtype=DataType.Float, is_cpu=False)
    T7 = fd.ops.reshape(T0, new_shape=[12, 128, 25, 32, 2])
    T8 = fd.ops.slice(T7, start_indices=[0, 0, 0, 0, 0], end_indices=[12, 128, 25, 32, 1], strides=[1, 1, 1, 1, 1])
    T9 = fd.ops.sum(T8, dims=[4], keepdim=False, dtype=DataType.Null)
    fd.add_output(T9)
```

### CODE REVIEW

**Comment by kevinstephano** [comment]
!build

**Comment by kevinstephano** [comment]
!build

**Comment by kevinstephano** [comment]
!build

**Comment by kevinstephano** [comment]
!build

**Review by jjsjann123 — COMMENTED** [review]
Looks great to me!

I'm wondering if we can converge `inline_def_` and `parent_` together as a single field. Looks like the two are almost always used as a pair. I'm wondering if it would be enough to have `inline_parent_` on `State`.

File: csrc/python_frontend/fusion_cache.cpp [hunk 1]
```diff
@@ -351,10 +351,12 @@ void FusionCache::print(std::ostream& os) const {
       }
 
       os << node->fusion_id << ":" << std::endl;
+      std::cout << "Fusion ID " << node->fusion_id << std::endl;
```
[the remark this answers is not in the document]
Comment (kevinstephano) [reply csrc/python_frontend/fusion_cache.cpp:354]: Nope, I left a comment in. :-(

File: csrc/python_frontend/fusion_record.h [hunk 2]
```diff
@@ -1850,18 +1899,50 @@ struct ScalarRecord : RecordFunctor {
...
+      } else {
+        os << "None";
+      }
 
-    os << ", dtype=" << dtypeToPyString(dtype_);
+      os << ", dtype=" << dtypeToPyString(dtype_);
 
-    if (close_function) {
-      os << ")";
+      if (close_function) {
```
[the remark this answers is not in the document]
Comment (kevinstephano) [reply csrc/python_frontend/fusion_record.h:1943]: I add this comment:

```
NOTE: close_function is not relevant for the inline definition as the
printing is specific to each operator and not partially done with the
base class print method.
```

File: csrc/python_frontend/fusion_state.cpp [hunk 3]
```diff
@@ -80,6 +105,22 @@ void FusionState::addRecord(RecordFunctor* record) {
   FUSER_PERF_SCOPE("FusionContainer::addRecord");
   recording_.emplace_back(record);
   num_recording_states_ += record->numOutputs();
+  RecordFunctor* state_record = recording_.back().get();
+  for (const auto& out : state_record->outputs()) {
+    if (out.index >= recording_state_.size()) {
+      // NOTE: This condition might occur during deserialization
+      recording_state_.resize(out.index + 1);
+      recording_state_.at(out.index) = out;
```
[the remark this answers is not in the document]
Comment (kevinstephano) [reply csrc/python_frontend/fusion_state.cpp]: I reworked this logic.  Inline definitions are not part of the caching, so it is not something that can be deserialized as deserialization only populates the cache and not a definition.

**Comment by kevinstephano** [comment]
!build

File: csrc/python_frontend/python_bindings.cpp [hunk 4]
```diff
@@ -107,7 +121,7 @@ Vector ShapeAsVector(ShapeType shape, FusionDefinition& fd) {
     // ```
     // would not work because the compiler would try to instantiate
     // define_vector_fn<Vector> and fail.
-    return define_vector_fn<ShapeType>(fd, shape);
+    return define_vector_fn<ShapeType>(fd, shape, true);
```
Comment (rdspring1) [inline csrc/python_frontend/python_bindings.cpp:124]:
```suggestion
    return define_vector_fn<ShapeType>(fd, shape, /*inline_def=*/true);
```

File: csrc/python_frontend/python_bindings.cpp [hunk 5]
```diff
@@ -85,7 +91,15 @@ Vector define_vector_fn(
...
-  return define_vector_base_fn(self, args);
+  return define_vector_base_fn(self, args, inline_def);
+}
+
+template <class ITERABLE>
+Vector define_vector_explicit_fn(
+    FusionDefinition& self,
+    ITERABLE& values,
+    PrimDataType dtype = DataType::Int) {
+  return define_vector_fn<ITERABLE>(self, values, false);
```
Comment (rdspring1) [inline csrc/python_frontend/python_bindings.cpp:102]:
```suggestion
  return define_vector_fn<ITERABLE>(self, values, /*inline_def=*/false);
```

File: csrc/python_frontend/python_bindings.cpp [hunk 6]
```diff
@@ -73,7 +76,10 @@ Vector define_vector_fn(
           " was neither symbolic(-1), zero_element(0), broadcast(1), or static(>1).");
       Scalar out = self.defineScalar();
       self.defineRecord(new ScalarRecord(
-          {self.recordingState(out())}, py::cast<int64_t>(item), dtype));
+          {self.recordingState(out())},
+          py::cast<int64_t>(item),
+          DataType::Int,
+          true));
```
Comment (rdspring1) [inline csrc/python_frontend/python_bindings.cpp:82]:
```suggestion
          /*inline_def=*/true));
```

**Review by jacobhinkle — APPROVED** [review]
LGTM

File: csrc/python_frontend/fusion_record.h [hunk 7]
```diff
@@ -1829,15 +1874,19 @@ struct ScalarRecord : RecordFunctor {
...
+            dtype_ == PrimDataType::Bool,
+            "A ScalarRecord for Bool inline definition not have a matching data type!");
         os << ((bool)value_ ? "True" : "False");
-      } else if (value_.is<std::complex<double>>()) {
-        os << std::showpoint << std::real(value_.as<std::complex<double>>())
-           << "+" << std::showpoint
-           << std::imag(value_.as<std::complex<double>>()) << "j";
       } else if (value_.is<double>()) {
+        NVF_CHECK(
+            dtype_ == PrimDataType::Double,
```
[the remark this answers is not in the document]
Comment (kevinstephano) [reply csrc/python_frontend/fusion_record.h:1888]: nvFuser can have other dtypes than just `Double`.  The default python floating point type is `double`.  Therefore, to specify anything other `double` the user needs to explicitly use `define_scalar()`.

File: csrc/python_frontend/fusion_record.h [hunk 8]
```diff
@@ -1850,18 +1899,53 @@ struct ScalarRecord : RecordFunctor {
           os << std::showpoint << value_.as<double>();
         }
       } else if (value_.is<int64_t>()) {
+        NVF_CHECK(
+            dtype_ == PrimDataType::Int,
```
[the remark this answers is not in the document]
Comment (kevinstephano) [reply csrc/python_frontend/fusion_record.h:1903]: nvFuser can have other dtypes than just `Int`.  The default python integer type is `int64_t`.  Therefore, to specify anything other `int64_t` the user needs to explicitly use `define_scalar()`.

File: csrc/python_frontend/fusion_record.h [hunk 9]
```diff
@@ -1850,18 +1899,53 @@ struct ScalarRecord : RecordFunctor {
...
-      os << "None";
-    }
+      RecordFunctor::print(os, false);
+      if (value_.hasValue()) {
+        if (value_.is<bool>()) {
+          os << ((bool)value_ ? "True" : "False");
+        } else if (value_.is<std::complex<double>>()) {
+          os << std::showpoint << std::real(value_.as<std::complex<double>>())
+             << "+" << std::showpoint
+             << std::imag(value_.as<std::complex<double>>()) << "j";
```
[the remark this answers is not in the document]
Comment (kevinstephano) [reply csrc/python_frontend/fusion_record.h:1922]: I am not sure why this was done for imaginary numbers as I didn't add this code.   I assume this is Xiang's code.

File: csrc/python_frontend/fusion_record.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/python_frontend/fusion_record.h:1888]: Do we actually have float scalar? I thought double is the only thing for scalar...

**Review by jjsjann123 — APPROVED** [review]
stamping since my comments has been resolved.

> I'm wondering if we can converge inline_def_ and parent_ together as a single field. Looks like the two are almost always used as a pair. I'm wondering if it would be enough to have inline_parent_ on State.

^^^ This I think is still a legit question. We don't have to do it in this PR, but I'm wondering if this is a good follow up?

File: csrc/python_frontend/fusion_record.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (jacobhinkle) [reply csrc/python_frontend/fusion_record.h:1888]: Right, right. Makes sense. We _do_ have scalars of other int and float types, but we can't _inline_ define them in Python.

**Comment by kevinstephano** [comment]
> stamping since my comments has been resolved.
> 
> > I'm wondering if we can converge inline_def_ and parent_ together as a single field. Looks like the two are almost always used as a pair. I'm wondering if it would be enough to have inline_parent_ on State.
> 
> ^^^ This I think is still a legit question. We don't have to do it in this PR, but I'm wondering if this is a good follow up?

I took your suggestion and modified the implementation.

**Comment by kevinstephano** [comment]
!build

File: csrc/python_frontend/fusion_state.h [hunk 10]
```diff
@@ -26,6 +37,10 @@ struct State {
   size_t index;
   //! StateType is either: Tensor, Scalar, or Vector
   serde::StateType stype;
+
+ private:
+  // This data member is only set if this state is inline defined!
+  std::optional<const RecordFunctor*> inline_def_record_;
```
Comment (jjsjann123) [inline csrc/python_frontend/fusion_state.h:43]: :clap:

**Pull request merged**